File: fragments/apis/fastify/datastore/postgres.ts
import { type FastifyInstance, type FastifyPluginOptions } from "fastify";
import {
  addItem,
  type CryptoCoin,
  getAllItems,
  getItemByTicker,
  getItemsAfterLaunchYear,
  removeItem,
  updateItem,
} from "../../../postgres_db/postgres_crud";

export const routes = (fastify: FastifyInstance, _: FastifyPluginOptions) => {
  fastify.get<{
    Reply: {
      200: CryptoCoin[];
      500: { error: string };
    };
  }>("/coins", async (request, reply) => {
    try {
      const items = await getAllItems();
      return items;
    } catch (error) {
      request.log.error(error, "Error fetching coins");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.get<{
    Params: { ticker: string };
    Reply: {
      200: CryptoCoin;
      404: void;
      500: { error: string };
    };
  }>("/coins/:ticker", async (request, reply) => {
    try {
      const { ticker } = request.params;
      const item = await getItemByTicker(ticker.toUpperCase());

      if (item.length === 0) {
        return reply.code(404).send();
      }
      return item[0];
    } catch (error) {
      request.log.error(error, "Error fetching coin by ticker");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.get<{
    Params: { year: number };
    Reply: {
      200: CryptoCoin[];
      500: { error: string };
    };
  }>("/coins/after/:year", async (request, reply) => {
    try {
      const { year } = request.params;
      const items = await getItemsAfterLaunchYear(year);
      return items;
    } catch (error) {
      request.log.error(error, "Error fetching coins after launch year");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.put<{
    Body: Omit<CryptoCoin, "id" | "ticker">;
    Params: { ticker: string };
    Reply: {
      200: void;
      400: { error: string };
      500: { error: string };
    };
  }>("/coins/:ticker", {
    schema: {
      body: {
        type: "object",
        required: ["name", "launched"],
        properties: {
          name: { type: "string" },
          launched: { type: "number" },
        },
      },
    },
  }, async (request, reply) => {
    try {
      const { ticker } = request.params;
      const { name, launched } = request.body;
      await addItem({
        name,
        launched,
        ticker: ticker.toUpperCase(),
      });
      return;
    } catch (error) {
      request.log.error(error, "Error adding coin");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.delete<{
    Params: { ticker: string };
    Reply: {
      204: void;
      500: { error: string };
    };
  }>("/coins/:ticker", async (request, reply) => {
    try {
      const { ticker } = request.params;
      await removeItem(ticker.toUpperCase());
      return reply.code(204).send();
    } catch (error) {
      request.log.error(error, "Error removing coin");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.patch<{
    Body: Omit<CryptoCoin, "id" | "ticker">;
    Params: { ticker: string };
    Reply: {
      200: CryptoCoin;
      404: void;
      500: { error: string };
    };
  }>("/coins/:ticker", {
    schema: {
      body: {
        type: "object",
        required: ["name", "launched"],
        properties: {
          name: { type: "string" },
          launched: { type: "number" },
        },
      },
    },
  }, async (request, reply) => {
    const { ticker } = request.params;
    const { name, launched } = request.body;
    try {
      const updatedItems = await updateItem({
        ticker: ticker.toUpperCase(),
        name,
        launched,
      });

      if (updatedItems.length === 0) {
        return reply.code(404).send();
      }
      return updatedItems[0];
    } catch (error) {
      request.log.error(error, "Error updating coin");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });
};

File: fragments/apis/fastapi/datastore/postgres.py
from typing import TypedDict, List
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from fragments.postgres_db.postgres_crud import (
    get_all_items,
    get_item_by_ticker,
    get_items_after_launch_year,
    add_item,
    remove_item,
    update_item,
)

postgres_router = APIRouter()


class CryptoCoin(TypedDict):
    id: int
    ticker: str
    name: str
    launched: int


class CoinBody(BaseModel):
    name: str
    launched: int


def tuple_to_coin(coin_tuple: tuple) -> CryptoCoin:
    return CryptoCoin(
        id=coin_tuple[0],
        ticker=coin_tuple[1],
        name=coin_tuple[2],
        launched=coin_tuple[3],
    )


@postgres_router.get("/coins", status_code=200)
async def get_coins() -> List[CryptoCoin]:
    items = get_all_items()
    return [tuple_to_coin(item) for item in items]


@postgres_router.get("/coins/after/{year}", status_code=200)
async def get_coins_after_year(year: int) -> List[CryptoCoin]:
    items = get_items_after_launch_year(year)
    return [tuple_to_coin(item) for item in items]


@postgres_router.get("/coins/{ticker}", status_code=200)
async def get_coin_by_ticker_route(ticker: str) -> CryptoCoin:
    item = get_item_by_ticker(ticker.upper())
    if item is None:
        raise HTTPException(status_code=404)
    return tuple_to_coin(item)


@postgres_router.put("/coins/{ticker}", status_code=200)
async def add_coin(ticker: str, body: CoinBody):
    add_item((ticker.upper(), body.name, body.launched))
    return None


@postgres_router.delete("/coins/{ticker}", status_code=204)
async def delete_coin(ticker: str):
    remove_item(ticker.upper())
    return None


@postgres_router.patch("/coins/{ticker}", status_code=200)
async def update_coin_route(ticker: str, body: CoinBody) -> CryptoCoin:
    result = update_item((ticker.upper(), body.name, body.launched))
    if result is None:
        raise HTTPException(status_code=404)
    return tuple_to_coin(result)
